File: tripsage/tools/schemas/calendar.py
"""
Google Calendar related schemas for TripSage tools.

This module provides Pydantic models for validating Google Calendar related data
when interacting with the Google Calendar MCP server.
"""

from enum import Enum
from typing import Any, Dict, List, Optional

from pydantic import Field, HttpUrl, model_validator

from tripsage.models.base import TripSageBaseResponse, TripSageModel


class EventTime(TripSageModel):
    """Model for event time information."""

    date_time: Optional[str] = Field(
        None, description="Date and time in ISO 8601 format"
    )
    date: Optional[str] = Field(None, description="Date in YYYY-MM-DD format")
    time_zone: Optional[str] = Field(None, description="Time zone name (IANA format)")

    @model_validator(mode="after")
    def validate_date_or_date_time(self):
        """Ensure either date or date_time is present, but not both."""
        if self.date_time is None and self.date is None:
            raise ValueError("Either date_time or date must be provided")
        if self.date_time is not None and self.date is not None:
            raise ValueError("Only one of date_time or date should be provided")
        return self


class EventStatus(str, Enum):
    """Event status options."""

    CONFIRMED = "confirmed"
    TENTATIVE = "tentative"
    CANCELLED = "cancelled"


class EventAttendee(TripSageModel):
    """Model for event attendee information."""

    email: str = Field(..., description="Email address of the attendee")
    display_name: Optional[str] = Field(
        None, description="Display name of the attendee"
    )
    response_status: Optional[str] = Field(
        None,
        description=(
            "Attendee's response status (accepted, declined, needsAction, tentative)"
        ),
    )
    optional: Optional[bool] = Field(None, description="Whether attendee is optional")


class EventReminder(TripSageModel):
    """Model for event reminder information."""

    method: str = Field(..., description="Reminder method (email, popup)")
    minutes: int = Field(..., description="Minutes before event to trigger reminder")


class EventVisibility(str, Enum):
    """Event visibility options."""

    DEFAULT = "default"
    PUBLIC = "public"
    PRIVATE = "private"
    CONFIDENTIAL = "confidential"


class Calendar(TripSageModel):
    """Model for Google Calendar information."""

    id: str = Field(..., description="Unique identifier for the calendar")
    summary: str = Field(..., description="Title of the calendar")
    description: Optional[str] = Field(None, description="Description of the calendar")
    time_zone: Optional[str] = Field(None, description="Time zone of the calendar")
    primary: Optional[bool] = Field(
        None, description="Whether this is the primary calendar"
    )


class Event(TripSageModel):
    """Model for Google Calendar event information."""

    id: Optional[str] = Field(None, description="Unique identifier for the event")
    calendar_id: Optional[str] = Field(
        None, description="ID of the calendar containing the event"
    )
    summary: str = Field(..., description="Title of the event")
    description: Optional[str] = Field(None, description="Description of the event")
    location: Optional[str] = Field(None, description="Location of the event")
    start: EventTime = Field(..., description="Start time information")
    end: EventTime = Field(..., description="End time information")
    status: Optional[EventStatus] = Field(None, description="Status of the event")
    html_link: Optional[HttpUrl] = Field(
        None, description="Link to the event in Google Calendar"
    )
    created: Optional[str] = Field(None, description="Creation time of the event")
    updated: Optional[str] = Field(None, description="Last update time of the event")
    attendees: Optional[List[EventAttendee]] = Field(
        None, description="List of attendees"
    )
    reminders: Optional[Dict[str, List[EventReminder]]] = Field(
        None, description="Reminders for the event"
    )
    recurrence: Optional[List[str]] = Field(
        None, description="Recurrence rules for the event"
    )
    visibility: Optional[EventVisibility] = Field(
        None, description="Visibility of the event"
    )
    conference_data: Optional[Dict] = Field(
        None, description="Conference data for the event"
    )


# API Request/Response Models
class ListCalendarsParams(TripSageModel):
    """Parameters for listing calendars."""

    max_results: Optional[int] = Field(
        None, description="Maximum number of calendars to return"
    )


class CalendarListResponse(TripSageBaseResponse):
    """Response model for listing calendars."""

    calendars: List[Calendar] = Field(..., description="List of calendars")
    next_page_token: Optional[str] = Field(None, description="Token for pagination")


class ListEventsParams(TripSageModel):
    """Parameters for listing events."""

    calendar_id: str = Field(..., description="ID of the calendar to list events from")
    time_min: Optional[str] = Field(None, description="Start time in ISO 8601 format")
    time_max: Optional[str] = Field(None, description="End time in ISO 8601 format")
    max_results: Optional[int] = Field(
        None, description="Maximum number of events to return"
    )
    single_events: Optional[bool] = Field(
        None, description="Whether to expand recurring events"
    )
    order_by: Optional[str] = Field(
        None, description="Order of events (startTime, updated)"
    )


class EventListResponse(TripSageBaseResponse):
    """Response model for listing events."""

    events: List[Event] = Field(..., description="List of events")
    next_page_token: Optional[str] = Field(None, description="Token for pagination")


class SearchEventsParams(TripSageModel):
    """Parameters for searching events."""

    calendar_id: str = Field(..., description="ID of the calendar to search")
    query: str = Field(..., description="Search query string")
    time_min: Optional[str] = Field(None, description="Start time in ISO 8601 format")
    time_max: Optional[str] = Field(None, description="End time in ISO 8601 format")
    max_results: Optional[int] = Field(
        None, description="Maximum number of events to return"
    )


class EventSearchResponse(TripSageBaseResponse):
    """Response model for searching events."""

    events: List[Event] = Field(..., description="List of matching events")
    next_page_token: Optional[str] = Field(None, description="Token for pagination")


class CreateEventParams(TripSageModel):
    """Parameters for creating an event."""

    calendar_id: str = Field(..., description="ID of the calendar to create event in")
    summary: str = Field(..., description="Title of the event")
    description: Optional[str] = Field(None, description="Description of the event")
    location: Optional[str] = Field(None, description="Location of the event")
    start: EventTime = Field(..., description="Start time information")
    end: EventTime = Field(..., description="End time information")
    attendees: Optional[List[EventAttendee]] = Field(
        None, description="List of attendees"
    )
    reminders: Optional[Dict[str, List[EventReminder]]] = Field(
        None, description="Reminders for the event"
    )
    recurrence: Optional[List[str]] = Field(
        None, description="Recurrence rules for the event"
    )
    visibility: Optional[EventVisibility] = Field(
        None, description="Visibility of the event"
    )
    conference_data: Optional[Dict] = Field(
        None, description="Conference data for the event"
    )


class UpdateEventParams(TripSageModel):
    """Parameters for updating an event."""

    calendar_id: str = Field(..., description="ID of the calendar containing the event")
    event_id: str = Field(..., description="ID of the event to update")
    summary: Optional[str] = Field(None, description="New title of the event")
    description: Optional[str] = Field(None, description="New description of the event")
    location: Optional[str] = Field(None, description="New location of the event")
    start: Optional[EventTime] = Field(None, description="New start time information")
    end: Optional[EventTime] = Field(None, description="New end time information")
    attendees: Optional[List[EventAttendee]] = Field(
        None, description="New list of attendees"
    )
    reminders: Optional[Dict[str, List[EventReminder]]] = Field(
        None, description="New reminders for the event"
    )
    recurrence: Optional[List[str]] = Field(None, description="New recurrence rules")
    visibility: Optional[EventVisibility] = Field(None, description="New visibility")


class DeleteEventParams(TripSageModel):
    """Parameters for deleting an event."""

    calendar_id: str = Field(..., description="ID of the calendar containing the event")
    event_id: str = Field(..., description="ID of the event to delete")


class ItineraryItemType(str, Enum):
    """Type of itinerary item."""

    FLIGHT = "flight"
    ACCOMMODATION = "accommodation"
    ACTIVITY = "activity"
    TRANSPORTATION = "transportation"
    DINING = "dining"
    OTHER = "other"


class ItineraryItem(TripSageModel):
    """Model for a TripSage itinerary item."""

    type: ItineraryItemType = Field(..., description="Type of itinerary item")
    title: str = Field(..., description="Title of the itinerary item")
    description: Optional[str] = Field(
        None, description="Description of the itinerary item"
    )
    location: Optional[str] = Field(None, description="Location of the itinerary item")
    start_time: str = Field(..., description="Start time in ISO 8601 format")
    end_time: Optional[str] = Field(None, description="End time in ISO 8601 format")
    duration_minutes: Optional[int] = Field(
        None, description="Duration in minutes (used if end_time not provided)"
    )
    time_zone: Optional[str] = Field(None, description="Time zone (IANA format)")
    confirmation_number: Optional[str] = Field(
        None, description="Confirmation or booking number"
    )
    details: Optional[Dict[str, Any]] = Field(None, description="Additional details")

    @model_validator(mode="after")
    def validate_time_info(self):
        """Ensure either end_time or duration_minutes is provided."""
        if not self.end_time and self.duration_minutes is None:
            raise ValueError("Either end_time or duration_minutes must be provided")
        return self


class CreateItineraryEventsParams(TripSageModel):
    """Parameters for creating events from an itinerary."""

    calendar_id: str = Field(..., description="ID of the calendar to create events in")
    itinerary_items: List[ItineraryItem] = Field(
        ..., description="List of itinerary items to convert to events"
    )
    trip_name: Optional[str] = Field(
        None, description="Name of the trip for event grouping"
    )


class CreateItineraryEventsResponse(TripSageBaseResponse):
    """Response for creating events from an itinerary."""

    created_events: List[Event] = Field(..., description="List of created events")
    failed_items: List[Dict[str, Any]] = Field(
        ..., description="List of items that failed to create with errors"
    )
    trip_name: Optional[str] = Field(None, description="Name of the trip")

File: frontend/src/domain/schemas/calendar.ts
/**
 * @fileoverview Canonical Zod v4 schemas for calendar entities, Google Calendar API
 * integration, and calendar tool input validation.
 *
 * Core schemas: Calendar entities, events, API responses
 * Tool schemas: Input validation for calendar tools (create events, export ICS)
 */

import { z } from "zod";
import { primitiveSchemas } from "./registry";

// ===== CORE SCHEMAS =====
// Core business logic schemas for calendar management

/** Zod schema for Google Calendar event visibility levels. */
export const eventVisibilitySchema = z.enum([
  "default",
  "public",
  "private",
  "confidential",
]);
/** TypeScript type for event visibility. */
export type EventVisibility = z.infer<typeof eventVisibilitySchema>;

/** Zod schema for Google Calendar event status values. */
export const eventStatusSchema = z.enum(["confirmed", "tentative", "cancelled"]);
/** TypeScript type for event status. */
export type EventStatus = z.infer<typeof eventStatusSchema>;

/** Zod schema for attendee response status values. */
export const attendeeResponseStatusSchema = z.enum([
  "needsAction",
  "declined",
  "tentative",
  "accepted",
]);
/** TypeScript type for attendee response status. */
export type AttendeeResponseStatus = z.infer<typeof attendeeResponseStatusSchema>;

/** Zod schema for reminder notification methods. */
export const reminderMethodSchema = z.enum(["email", "popup", "sms"]);
/** TypeScript type for reminder methods. */
export type ReminderMethod = z.infer<typeof reminderMethodSchema>;

/** Zod schema for Google Calendar event date/time with timezone support. */
export const eventDateTimeSchema = z.object({
  date: z
    .string()
    .regex(/^\d{4}-\d{2}-\d{2}$/, { error: "Date must be in YYYY-MM-DD format" })
    .optional()
    .or(z.date().transform((d) => d.toISOString().split("T")[0])),
  dateTime: z
    .union([z.date(), z.iso.datetime()])
    .optional()
    .transform((val) => (typeof val === "string" ? new Date(val) : val)),
  timeZone: z.string().optional(),
});
/** TypeScript type for event date/time. */
export type EventDateTime = z.infer<typeof eventDateTimeSchema>;

/** Zod schema for Google Calendar event reminders. */
export const eventReminderSchema = z.object({
  method: reminderMethodSchema.default("popup"),
  minutes: z.number().int().min(0).max(40320),
});
/** TypeScript type for event reminders. */
export type EventReminder = z.infer<typeof eventReminderSchema>;

/** Zod schema for Google Calendar event attendees. */
export const eventAttendeeSchema = z.object({
  additionalGuests: z.number().int().min(0).default(0),
  comment: z.string().optional(),
  displayName: z.string().optional(),
  email: primitiveSchemas.email,
  optional: z.boolean().default(false),
  responseStatus: attendeeResponseStatusSchema.default("needsAction"),
});
/** TypeScript type for event attendees. */
export type EventAttendee = z.infer<typeof eventAttendeeSchema>;

/** Zod schema for Google Calendar conference data. */
export const conferenceDataSchema = z.object({
  conferenceId: z.string().optional(),
  conferenceSolution: z.record(z.string(), z.unknown()).optional(),
  entryPoints: z.array(z.record(z.string(), z.unknown())).optional(),
  notes: z.string().optional(),
});
/** TypeScript type for conference data. */
export type ConferenceData = z.infer<typeof conferenceDataSchema>;

/** Zod schema for Google Calendar extended properties. */
export const extendedPropertiesSchema = z.object({
  private: z.record(z.string(), z.string()).default({}),
  shared: z.record(z.string(), z.string()).default({}),
});
/** TypeScript type for extended properties. */
export type ExtendedProperties = z.infer<typeof extendedPropertiesSchema>;

/**
 * Zod schema for Google Calendar events.
 */
export const calendarEventSchema = z.object({
  attendees: z.array(eventAttendeeSchema).default([]),
  attendeesOmitted: z.boolean().default(false),
  colorId: z.string().optional(),
  conferenceData: conferenceDataSchema.optional(),
  created: z.date().optional(),
  creator: z.record(z.string(), z.unknown()).optional(),
  description: z.string().max(8192).optional(),
  end: eventDateTimeSchema,
  endTimeUnspecified: z.boolean().default(false),
  etag: z.string().optional(),
  extendedProperties: extendedPropertiesSchema.optional(),
  hangoutLink: primitiveSchemas.url.optional(),
  htmlLink: primitiveSchemas.url.optional(),
  iCalUID: z.string().optional(),
  id: z.string().optional(),
  location: z.string().max(1024).optional(),
  organizer: z.record(z.string(), z.unknown()).optional(),
  originalStartTime: eventDateTimeSchema.optional(),
  recurrence: z.array(z.string()).optional(),
  recurringEventId: z.string().optional(),
  reminders: z
    .object({
      overrides: z.array(eventReminderSchema).optional(),
      useDefault: z.boolean().optional(),
    })
    .default({ useDefault: true }),
  sequence: z.number().int().min(0).default(0),
  start: eventDateTimeSchema,
  status: eventStatusSchema.default("confirmed"),
  summary: z.string().min(1).max(1024),
  transparency: z.enum(["opaque", "transparent"]).default("opaque"),
  travelMetadata: z.record(z.string(), z.unknown()).optional(),
  updated: z.date().optional(),
  visibility: eventVisibilitySchema.default("default"),
});
/** TypeScript type for calendar events. */
export type CalendarEvent = z.infer<typeof calendarEventSchema>;

// ===== API SCHEMAS =====
// Request/response schemas for calendar API endpoints

/**
 * Zod schema for creating new Google Calendar events.
 * Validates event parameters including dates, attendees, and reminders.
 */
export const createEventRequestSchema = z.object({
  attendees: z.array(eventAttendeeSchema).default([]),
  conferenceDataVersion: z.number().int().optional(),
  description: z.string().max(8192).optional(),
  end: eventDateTimeSchema,
  location: z.string().max(1024).optional(),
  recurrence: z.array(z.string()).optional(),
  reminders: z
    .object({
      overrides: z.array(eventReminderSchema).optional(),
      useDefault: z.boolean().optional(),
    })
    .optional(),
  start: eventDateTimeSchema,
  summary: z.string().min(1).max(1024),
  timeZone: z.string().optional(),
  transparency: z.enum(["opaque", "transparent"]).default("opaque"),
  travelMetadata: z.record(z.string(), z.unknown()).optional(),
  visibility: eventVisibilitySchema.default("default"),
});
/** TypeScript type for create event requests. */
export type CreateEventRequest = z.infer<typeof createEventRequestSchema>;

/** Zod schema for updating existing Google Calendar events. */
export const updateEventRequestSchema = z.object({
  attendees: z.array(eventAttendeeSchema).optional(),
  description: z.string().max(8192).optional(),
  end: eventDateTimeSchema.optional(),
  location: z.string().max(1024).optional(),
  recurrence: z.array(z.string()).optional(),
  reminders: z
    .object({
      overrides: z.array(eventReminderSchema).optional(),
      useDefault: z.boolean().optional(),
    })
    .optional(),
  start: eventDateTimeSchema.optional(),
  summary: z.string().min(1).max(1024).optional(),
  timeZone: z.string().optional(),
  transparency: z.enum(["opaque", "transparent"]).optional(),
  travelMetadata: z.record(z.string(), z.unknown()).optional(),
  visibility: eventVisibilitySchema.optional(),
});
/** TypeScript type for update event requests. */
export type UpdateEventRequest = z.infer<typeof updateEventRequestSchema>;

/** Zod schema for Google Calendar list entries. */
export const calendarListEntrySchema = z.object({
  accessRole: z.string().default("reader"),
  backgroundColor: z.string().optional(),
  colorId: z.string().optional(),
  conferenceProperties: z.record(z.string(), z.unknown()).optional(),
  defaultReminders: z.array(eventReminderSchema).default([]),
  deleted: z.boolean().default(false),
  description: z.string().optional(),
  etag: z.string().optional(),
  foregroundColor: z.string().optional(),
  hidden: z.boolean().default(false),
  id: z.string(),
  kind: z.literal("calendar#calendarListEntry").default("calendar#calendarListEntry"),
  location: z.string().optional(),
  notificationSettings: z.record(z.string(), z.unknown()).optional(),
  primary: z.boolean().default(false),
  selected: z.boolean().default(true),
  summary: z.string(),
  summaryOverride: z.string().optional(),
  timeZone: z.string().optional(),
});
/** TypeScript type for calendar list entries. */
export type CalendarListEntry = z.infer<typeof calendarListEntrySchema>;

/** Zod schema for Google Calendar list responses. */
export const calendarListSchema = z.object({
  etag: z.string().optional(),
  items: z.array(calendarListEntrySchema).default([]),
  kind: z.literal("calendar#calendarList").default("calendar#calendarList"),
  nextPageToken: z.string().optional(),
  nextSyncToken: z.string().optional(),
});
/** TypeScript type for calendar list responses. */
export type CalendarList = z.infer<typeof calendarListSchema>;

/** Zod schema for free/busy calendar items. */
export const freeBusyCalendarItemSchema = z.object({
  id: z.string(),
});
/** TypeScript type for free/busy calendar items. */
export type FreeBusyCalendarItem = z.infer<typeof freeBusyCalendarItemSchema>;

/**
 * Zod schema for Google Calendar free/busy requests.
 * Validates parameters for checking calendar availability.
 */
export const freeBusyRequestSchema = z
  .object({
    calendarExpansionMax: z.number().int().min(1).max(50).default(50),
    groupExpansionMax: z.number().int().min(1).max(100).default(50),
    items: z.array(freeBusyCalendarItemSchema).min(1),
    timeMax: z.union([z.date(), z.string().transform((val) => new Date(val))]),
    timeMin: z.union([z.date(), z.string().transform((val) => new Date(val))]),
    timeZone: z.string().optional(),
  })
  .refine((data) => data.timeMax > data.timeMin, {
    error: "timeMax must be after timeMin",
    path: ["timeMax"],
  });
/** TypeScript type for free/busy requests. */
export type FreeBusyRequest = z.infer<typeof freeBusyRequestSchema>;

/** Zod schema for Google Calendar free/busy responses. */
export const freeBusyResponseSchema = z.object({
  calendars: z.record(z.string(), z.record(z.string(), z.unknown())).default({}),
  groups: z.record(z.string(), z.unknown()).default({}),
  kind: z.literal("calendar#freeBusy").default("calendar#freeBusy"),
  timeMax: z.date(),
  timeMin: z.date(),
});
/** TypeScript type for free/busy responses. */
export type FreeBusyResponse = z.infer<typeof freeBusyResponseSchema>;

/** Zod schema for Google Calendar events list requests. */
export const eventsListRequestSchema = z.object({
  alwaysIncludeEmail: z.boolean().default(false),
  calendarId: z.string().default("primary"),
  iCalUID: z.string().optional(),
  maxAttendees: z.number().int().min(1).optional(),
  maxResults: z.number().int().min(1).max(2500).default(250),
  orderBy: z.enum(["startTime", "updated"]).optional(),
  pageToken: z.string().optional(),
  privateExtendedProperty: z.array(z.string()).optional(),
  q: z.string().optional(),
  sharedExtendedProperty: z.array(z.string()).optional(),
  showDeleted: z.boolean().default(false),
  showHiddenInvitations: z.boolean().default(false),
  singleEvents: z.boolean().default(false),
  syncToken: z.string().optional(),
  timeMax: z.date().optional(),
  timeMin: z.date().optional(),
  timeZone: z.string().optional(),
  updatedMin: z.date().optional(),
});
/** TypeScript type for events list requests. */
export type EventsListRequest = z.infer<typeof eventsListRequestSchema>;

/** Zod schema for Google Calendar events list responses. */
export const eventsListResponseSchema = z.object({
  accessRole: z.string().optional(),
  defaultReminders: z.array(eventReminderSchema).default([]),
  description: z.string().optional(),
  etag: z.string().optional(),
  items: z.array(calendarEventSchema).default([]),
  kind: z.literal("calendar#events").default("calendar#events"),
  nextPageToken: z.string().optional(),
  nextSyncToken: z.string().optional(),
  summary: z.string().optional(),
  timeZone: z.string().optional(),
  updated: z.date().optional(),
});
/** TypeScript type for events list responses. */
export type EventsListResponse = z.infer<typeof eventsListResponseSchema>;

/**
 * Zod schema for POST /api/calendar/ics/import request body.
 * Validates ICS import request parameters for route handlers.
 */
export const icsImportRequestSchema = z.object({
  icsData: z.string().min(1, { error: "ICS data is required" }),
  validateOnly: z.boolean().default(true),
});

/** TypeScript type for ICS import requests. */
export type IcsImportRequest = z.infer<typeof icsImportRequestSchema>;

/**
 * Zod schema for POST /api/calendar/ics/export request body.
 * Validates ICS export request parameters for route handlers.
 */
export const icsExportRequestSchema = z.object({
  calendarName: z.string().default("TripSage Calendar"),
  events: z
    .array(calendarEventSchema)
    .min(1, { error: "At least one event is required" }),
  timezone: z.string().optional(),
});

/** TypeScript type for ICS export requests. */
export type IcsExportRequest = z.infer<typeof icsExportRequestSchema>;

/**
 * Calendar connection status response type.
 * Used by calendar status API and components.
 */
export interface CalendarStatusResponse {
  connected: boolean;
  calendars?: Array<{
    id: string;
    summary: string;
    description?: string;
    timeZone?: string;
    primary?: boolean;
    accessRole?: string;
  }>;
  message?: string;
}

// ===== TOOL INPUT SCHEMAS =====
// Schemas for calendar tool input validation and processing

/**
 * Schema for creating calendar events with Google Calendar.
 * Validates calendar event creation parameters for AI tools.
 */
export const createCalendarEventInputSchema = createEventRequestSchema.extend({
  calendarId: z
    .string()
    .default("primary")
    .describe("Google Calendar ID to create event in"),
});

/**
 * Schema for exporting calendar events to ICS format.
 * Validates itinerary export parameters for AI tools.
 */
export const exportItineraryToIcsInputSchema = z.strictObject({
  calendarName: z
    .string()
    .default("TripSage Itinerary")
    .describe("Name for the calendar in ICS export"),
  events: z
    .array(calendarEventSchema)
    .min(1, { error: "At least one event is required" })
    .describe("List of calendar events to export"),
  timezone: z.string().nullable().describe("Timezone for the exported calendar"),
});
